builtin/providers/azure/resource_azure_security_group_rule.go

```diff
@@ -0,0 +1,318 @@
+			},
+			"priority": &schema.Schema{
+				Type:        schema.TypeInt,
+				Required:    true,
+				Description: parameterDescriptions["netsecgroup_priority"],
+			},
+			"action": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: parameterDescriptions["netsecgroup_action"],
+			},
+			"source_address_prefix": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: parameterDescriptions["netsecgroup_src_addr_prefix"],
+			},
+			"source_port_range": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: parameterDescriptions["netsecgropu_src_port_range"],
+			},
+			"destination_address_prefix": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: parameterDescriptions["netsecgroup_dest_addr_prefix"],
+			},
+			"destination_port_range": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: parameterDescriptions["netsecgroup_dest_port_range"],
+			},
+			"protocol": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: parameterDescriptions["netsecgroup_protocol"],
+			},
+		},
+	}
+}
+
+// resourceAzureSecurityGroupRuleCreate does all the necessary API calls to
+// create a new network security group rule on Azure.
+func resourceAzureSecurityGroupRuleCreate(d *schema.ResourceData, meta interface{}) error {
+	azureClient := meta.(*AzureClient)
+	managementClient := azureClient.managementClient
+	netSecClient := netsecgroup.NewClient(managementClient)
+
+	// create and configure the RuleResponse:
+	rule := netsecgroup.RuleRequest{
+		// TODO(aznashwan): security checks here:
+		Name:                     d.Get("name").(string),
+		Type:                     netsecgroup.RuleType(d.Get("type").(string)),
+		Priority:                 d.Get("priority").(int),
+		Action:                   netsecgroup.RuleAction(d.Get("action").(string)),
+		SourceAddressPrefix:      d.Get("source_address_prefix").(string),
+		SourcePortRange:          d.Get("source_port_range").(string),
+		DestinationAddressPrefix: d.Get("destination_address_prefix").(string),
+		DestinationPortRange:     d.Get("destination_port_range").(string),
+		Protocol:                 netsecgroup.RuleProtocol(d.Get("protocol").(string)),
+	}
+
+	// send the create request to Azure:
+	log.Println("[INFO] Sending network security group rule creation request to Azure.")
+	reqID, err := netSecClient.SetNetworkSecurityGroupRule(
+		d.Get("security_group_name").(string),
+		rule,
+	)
+	if err != nil {
+		return fmt.Errorf("Error sending network security group rule creation request to Azure: %s", err)
+	}
+	err = managementClient.WaitForOperation(reqID, nil)
+	if err != nil {
+		return fmt.Errorf("Error creating network security group rule on Azure: %s", err)
+	}
+
+	d.SetId(getRandomStringLabel(50))
```

Comment:
Same question as above about generated id vs name - if we have to look it up by name already, why not just use that name as the internal id?